### PR TITLE
Authorization Code Flow for Single Page Applications: Pre-Alpha Preparation

### PR DESCRIPTION
**Pull request by pkanher617**
This PR fixes lint rules, adds developer comments and adds updated README documentation. This PR is the last before the release of alpha versions of the library for the browser and common packages.

### CODE REVIEW

File: lib/msal-browser/README.md [hunk 1]
```diff
@@ -10,12 +10,46 @@ The MSAL library for JavaScript enables client-side JavaScript applications to a
...
+### Build Library
+```javascript
+// To run build only for browser package
+npm run build
+// To run build for common and browser package
+npm run build:all
+```
+
+### Run
+To run sample, ensure the client_id and client_secret are updated and filled in.
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: We need to make it very clear that this is temporary (should not be provided once portal has been updated) and should not be deployed to production environments.

File: lib/msal-browser/README.md [hunk 2]
```diff
@@ -10,12 +10,46 @@ The MSAL library for JavaScript enables client-side JavaScript applications to a
 [![Build Status](https://travis-ci.org/AzureAD/microsoft-authentication-library-for-js.png?branch=dev)](https://travis-ci.org/AzureAD/microsoft-authentication-library-for-js)[![npm version](https://img.shields.io/npm/v/msal.svg?style=flat)](https://www.npmjs.com/package/msal)[![npm version](https://img.shields.io/npm/dm/msal.svg)](https://nodei.co/npm/msal/)
 
 ## Installation
-### Via NPM:
+### Via NPM (Not available yet):
 
     npm install msal-browser
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: `@azure/msal-browser`

File: lib/msal-browser/package.json [hunk 3]
```diff
@@ -10,7 +10,7 @@
         "type": "git",
         "url": "https://github.com/AzureAD/microsoft-authentication-library-for-js.git"
     },
-    "version": "0.0.1",
+    "version": "2.0.0-alpha",
```
Comment (jasonnutter) [inline lib/msal-browser/package.json]: `2.0.0-alpha.0` to be consistent with how we have been doing pre-release versions.

File: lib/msal-core/README.md [hunk 4]
```diff
@@ -221,24 +221,6 @@ See [Request and Response Data Types](https://github.com/AzureAD/microsoft-authe
...
-    };
-    var graphEndpoint = "https://graph.microsoft.com/v1.0/me";
-
-    fetch(graphEndpoint, options)
-        .then(resp => {
-             //do something with response
-        });
-```
-
-You can learn further details about MSAL.js functionality documented in the [MSAL Wiki](https://github.com/AzureAD/microsoft-authentication-library-for-js/wiki) and find complete [code samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/wiki/Samples).
```
Comment (jasonnutter) [inline lib/msal-core/README.md]: Why is this deleted? Still relevant for `msal-core`.

File: samples/VanillaJSTestApp/index_authCode.html [hunk 5]
```diff
@@ -24,6 +24,7 @@
     const msalConfig = {
         auth: {
             clientId: "Enter_your_client_id",
+            clientSecret: "Enter_your_client_secret",
```
Comment (jasonnutter) [inline samples/VanillaJSTestApp/index_authCode.html]: Again, make it clear (with comments) that this is only needed for pre-release versions.

File: lib/msal-common/src/logger/Logger.ts [hunk 6]
```diff
@@ -2,46 +2,52 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License.
  */
+// package.json
```
Comment (jasonnutter) [inline lib/msal-common/src/logger/Logger.ts]: A lot of these comments feel superfluous, and probably aren't necessary.

File: lib/msal-common/package.json [hunk 7]
```diff
@@ -10,7 +10,7 @@
         "type": "git",
         "url": "https://github.com/AzureAD/microsoft-authentication-library-for-js.git"
     },
-    "version": "0.0.1",
+    "version": "1.0.0-alpha",
```
Comment (jasonnutter) [inline lib/msal-common/package.json]: Same comment about `alpha.0`

File: lib/msal-browser/package.json [hunk 8]
```diff
@@ -10,7 +10,7 @@
...
     "description": "Microsoft Authentication Library for js",
     "keywords": [
         "implicit",
```
Comment (jasonnutter) [inline lib/msal-browser/package.json]: I don't see `@azure/msal-common` listed as a dependency.

File: lib/msal-browser/README.md [hunk 9]
```diff
@@ -38,14 +72,129 @@ We plan to continue support for the implicit flow in the library.
...
 Before using MSAL.js you will need to [register an application in Azure AD](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app) to get a valid `clientId` for configuration, and to register the routes that your app will accept redirect traffic on.
 
 > TBD
 
+#### 1. Initializing the Public Client Application
+
+`PublicClientApplication` can be configured with a variety of different options, detailed in our [Wiki](), but the only required parameter is `auth.clientId`.
+
+```javascript
+import * as msal from "msal";
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: `import * as msal from "@azure/msal-browser"`

File: lib/msal-browser/README.md [hunk 10]
```diff
@@ -38,14 +72,129 @@ We plan to continue support for the implicit flow in the library.
...
+};
+
+const msalInstance = new msal.PublicClientApplication(msalConfig);
+```
+
+#### 2. Choose interaction type - redirect or popup
+
+Choose which APIs you will use in your authentication flows:
+- loginRedirect and acquireTokenRedirect
+- loginPopup and acquireTokenPopup
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: Backticks around these

File: lib/msal-browser/README.md [hunk 11]
```diff
@@ -38,14 +72,129 @@ We plan to continue support for the implicit flow in the library.
...
+});
+```
+
+#### 3. Login the user
+
+Your app must login the user with either the `loginPopup` or the `loginRedirect` method to establish user context.
+
+When the login methods are called and the authentication of the user is completed by the Azure AD service, an [id token](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens) is returned which is used to identify the user with some basic information.
+
+When you login a user, you can pass in scopes that the user can pre consent to on login, however this is not required. Please note that consenting to scopes on login, does not return an access_token for these scopes, but gives you the opportunity to obtain a token silently with these scopes passed in, with no further interaction from the user.
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: "that the user can pre-consent to, however, this is not required"

File: lib/msal-browser/src/app/Configuration.ts [hunk 12]
```diff
@@ -2,14 +2,15 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License.
  */
+// Common package imports
 import { AuthOptions, SystemOptions, LoggerOptions, INetworkModule, LogLevel } from "msal-common";
+// Utils
 import { BrowserUtils } from "../utils/BrowserUtils";
+// Constants
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/Configuration.ts]: I would argue these sort of comments aren't needed.

File: lib/msal-browser/src/app/Configuration.ts [hunk 13]
```diff
@@ -54,20 +56,24 @@ export type Configuration = {
     system?: BrowserSystemOptions
 };
 
+// Default auth options for browser
 const DEFAULT_AUTH_OPTIONS: BrowserAuthOptions = {
     clientId: "",
+    clientSecret: "",
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/Configuration.ts]: I would even consider renaming this to something like `_temp_clientSecret` to make it very clear this is temporary.

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 14]
```diff
@@ -2,30 +2,26 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License.
  */
-import { AuthError, Account, AuthResponse, AuthorizationCodeModule, AuthenticationParameters, INetworkModule, TokenResponse, UrlString, TemporaryCacheKeys, TokenRenewParameters } from "msal-common";
-import { BrowserStorage } from "../cache/BrowserStorage";
+// Common package imports
+import { Account, AuthorizationCodeModule, AuthenticationParameters, INetworkModule, TokenResponse, UrlString, TemporaryCacheKeys, TokenRenewParameters } from "msal-common";
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/PublicClientApplication.ts]: These references need to be updated to `@azure/msal-common`

File: lib/msal-browser/src/cache/BrowserStorage.ts [hunk 15]
```diff
@@ -2,32 +2,54 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License.
  */
-import { ICacheStorage, Constants, PersistentCacheKeys, TemporaryCacheKeys, ErrorCacheKeys } from "msal-common";
+// Common package imports
+import { ICacheStorage, Constants, PersistentCacheKeys, TemporaryCacheKeys } from "msal-common";
```
Comment (jasonnutter) [inline lib/msal-browser/src/cache/BrowserStorage.ts]: `@azure/msal-common`

File: lib/msal-browser/README.md [hunk 16]
```diff
@@ -38,14 +72,129 @@ We plan to continue support for the implicit flow in the library.
...
+
+You can use `acquireTokenRedirect` or `acquireTokenPopup` to initiate interactive requests, although, it is best practice to only show interactive experiences if you are unable to obtain a token silently due to interaction required errors. If you are using an interactive token call, it must match the login method used in your application. (`loginPopup`=> `acquireTokenPopup`, `loginRedirect` => `acquireTokenRedirect`).
+
+If the `acquireTokenSilent` call fails, you may need to initiate an interactive request. This could happen for many reasons including scopes that have been revoked, expired tokens, or password changes.
+
+`acquireTokenSilent` will look for a valid token in the cache, and if it is close to expiring or does not exist, will automatically try to refresh it for you.
+
+See [Request and Response Data Types]() for reference.
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: Missing link.

File: lib/msal-browser/README.md [hunk 17]
```diff
@@ -38,14 +72,129 @@ We plan to continue support for the implicit flow in the library.
...
+        msalInstance.acquireTokenSilent(tokenRequest)
+            .then(response => {
+                // get access token from response
+                // response.accessToken
+            })
+            .catch(err => {
+                // handle error
+            });
+    } else {
+        // user is not logged in, you will need to log them in to acquire a token
```
Comment (jasonnutter) [inline lib/msal-browser/README.md]: Could have a sample that calls one of the login methods.

File: lib/msal-core/README.md (hunk 4 above)
Comment (pkanher617) [reply lib/msal-core/README.md]: duplicate sections. Look at the entire file.

File: samples/VanillaJSTestApp/index_authCode.html (hunk 5 above)
Comment (pkanher617) [reply samples/VanillaJSTestApp/index_authCode.html]: Done, changed the name of the variable, added logger warnings, added comments in the readme.

File: lib/msal-browser/README.md (hunk 1 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: Done, changed the name of the variable, added logger warnings, added comments in the readme.

File: lib/msal-browser/README.md (hunk 2 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: done.

File: lib/msal-browser/package.json (hunk 3 above)
Comment (pkanher617) [reply lib/msal-browser/package.json]: done.

File: lib/msal-common/package.json (hunk 7 above)
Comment (pkanher617) [reply lib/msal-common/package.json]: done.

File: lib/msal-browser/package.json (hunk 8 above)
Comment (pkanher617) [reply lib/msal-browser/package.json]: added dependency.

File: lib/msal-browser/README.md (hunk 9 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: done.

File: lib/msal-browser/README.md (hunk 11 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: done.

File: lib/msal-browser/README.md (hunk 10 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: done.

File: lib/msal-browser/src/app/Configuration.ts (hunk 13 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/Configuration.ts]: done

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 14 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: done.

File: lib/msal-browser/src/cache/BrowserStorage.ts (hunk 15 above)
Comment (pkanher617) [reply lib/msal-browser/src/cache/BrowserStorage.ts]: done

File: lib/msal-browser/README.md (hunk 16 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: Links will be added later, going to do some updates to the wiki.

File: lib/msal-browser/README.md (hunk 17 above)
Comment (pkanher617) [reply lib/msal-browser/README.md]: We don't throw interaction required anymore, need to do more research to understand how this will work.

File: lib/msal-browser/README.md [hunk 18]
```diff
@@ -10,12 +10,46 @@ The MSAL library for JavaScript enables client-side JavaScript applications to a
...
+cd samples/VanillaJSTestApp
+npm start
+```
+
+## How to Run Samples for Authorization Code Flow (Alpha)
+The current VanillaJSTestApp sample is set up to run the authorization code flow in the browser. However, there are a few pre-requisites that you will need to complete before being able to run the VanillaJS sample. 
+
+### Pre-requisites
+
+1. Create an application registration in the portal. Use whatever audience you wish, as long as it is testable on your machine. It is recommended to use the common audience (a.k.a. accounts in any Azure tenant) for simplest use.
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-browser/README.md]: Yes I will have more info about portal registration. There is a Microsoft Docs page that describes the process, I will link it here.

File: lib/msal-browser/README.md [hunk 19]
```diff
@@ -10,12 +10,46 @@ The MSAL library for JavaScript enables client-side JavaScript applications to a
...
+    - Go the Certificates & Secrets tab. Create a Client Secret that you can use for testing. This will most likely not be a required step in future production versions for Auth Code in the browser, but for now in order to run the alpha you will need to create one.
+2. Keep the app registration page open. You will now need a browser with CORS disabled in order to be able to retrieve tokens from the token endpoint. This is once again not a recommended production setting, but for the purposes of this alpha you should follow these instructions:
+    - We recommend using Chrome for this. You can follow the steps [here](https://alfilatov.com/posts/run-chrome-without-cors/) to figure out how to run Chrome withput CORS enabled for your OS.
+    - For Windows machines, you can do the following:
+        - Right click on your desktop -> New -> Shortcut
+        - Paste the following: 
+        ```javascript
+        "C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" --disable-web-security --disable-gpu --user-data-dir=~/chromeTemp
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-browser/README.md]: The link given describes how to use the script in most environments. I updated the README to point people not using windows in the right direction.

File: lib/msal-browser/README.md (hunk 17 above)
Comment (jasonnutter) [reply lib/msal-browser/README.md]: Ah, I'm guessing this will be status code based now.

File: lib/msal-browser/package.json [hunk 20]
```diff
@@ -85,5 +85,7 @@
         "tslint": "^5.20.0",
         "typescript": "^3.6.4"
     },
-    "dependencies": {}
+    "dependencies": {
+        "@azure/msal-common": "1.0.0-alpha"
```
Comment (jasonnutter) [inline lib/msal-browser/package.json]: `1.0.0-alpha.0`

File: samples/VanillaJSTestApp/index_authCode.html [hunk 21]
```diff
@@ -24,6 +24,7 @@
     const msalConfig = {
         auth: {
             clientId: "Enter_your_client_id",
+            tmp_clientSecret: "Enter_your_client_id",
```
Comment (jasonnutter) [inline samples/VanillaJSTestApp/index_authCode.html]: Secret

File: lib/msal-core/README.md (hunk 4 above)
Comment (jasonnutter) [reply lib/msal-core/README.md]: Indeed, looked in the wrong direction.

File: lib/msal-browser/package.json (hunk 20 above)
Comment (pkanher617) [reply lib/msal-browser/package.json]: done.
